### PR TITLE
[Impeller] A toolkit for managed handles to Android NDK vended objects.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Only available on Android device API levels >= 29. Proc table is setup has versioning checks. All handles are type safe. Collection of handles takes into account cleanup tasks (like reparenting surface controls). The proc table contains code duplicated in ndk_helpers and I will remove that in favor of this in a subsequent patch.

Part of https://github.com/flutter/engine/pull/51213 being chopped up.

### CODE REVIEW

File: impeller/toolkit/android/hardware_buffer.cc
Comment (jonahwilliams) [inline impeller/toolkit/android/hardware_buffer.cc:82]: This seems like it should be an error state (std::option or statusOr) instead of returning a size that wasn't requested.

File: impeller/toolkit/android/hardware_buffer.h
Comment (jonahwilliams) [inline impeller/toolkit/android/hardware_buffer.h:22]: 😆

File: impeller/toolkit/android/hardware_buffer.h
Comment (jonahwilliams) [inline impeller/toolkit/android/hardware_buffer.h:52]: Note size requirements here.

File: impeller/toolkit/android/native_window.h
Comment (jonahwilliams) [inline impeller/toolkit/android/native_window.h:33]: current size?

File: impeller/toolkit/android/surface_control.h
Comment (jonahwilliams) [inline impeller/toolkit/android/surface_control.h:22]: Does the surface control have any threading restrictions? In general, if we're aware of any threading restrictions for parts of these APIs it would be good to note them.

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (jonahwilliams) [inline impeller/toolkit/android/toolkit_android_unittests.cc:21]: Given that we just went through a fire drill where tests weren't actually running (because they failed unsafely) - how can we guarantee that we get a failure somewhere if we accidentally stop running this on API 29+ ?

**Review by jonahwilliams — COMMENTED** [review]
Do we intented to (eventually) replace the existing AHardwareBuffer and NDK wrappers used in the android embedder with this entirely?

File: impeller/toolkit/android/README.md
Comment (dnfield) [inline impeller/toolkit/android/README.md:4]: Nit: please wrap at 80 chars.

Consider explaining that this works by dynamic runtime lookup here, and why (so that you can build targetting older API levels while still taking advantage of newer features)

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (dnfield) [inline impeller/toolkit/android/toolkit_android_unittests.cc:20]: nit: This might read a little clearer and avoid a builtin via

```suggestion
    if (android_get_device_api_level() < 29) {
```

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (dnfield) [reply impeller/toolkit/android/toolkit_android_unittests.cc:21]: We have the same problem with the existing ndk_helpers_unittests.

We do, in fact, currently run emulators that will run this test, but we should probably patch ... something to fail if this test never runs at all anywhere.

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (dnfield) [inline impeller/toolkit/android/toolkit_android_unittests.cc:14]: I don't know if you want to do this as a follow up, but related to Jonah's question:

The tests here don't have as much coverage as the ndk_helpers_unittests (specifcially, there aren't assertions that the API works and that we can allocate/destroy these objects or that the API symbols are correctly resolved). Maybe this will be good for a follow up patch, but consider porting those tests to this toolkit, deleting the FML versions, and porting the Android embedding to use this toolkit?

If that's not possible, I'd rather not land this in Impeller - I'm somewhat concerned about having more places that we do these lookups and introduce typos/untested code, especially since we just got that cleaned up.

Comment (matanlurey) [reply impeller/toolkit/android/toolkit_android_unittests.cc:14]: > If that's not possible, I'd rather not land this in Impeller - I'm somewhat concerned about having more places that we do these lookups and introduce typos/untested code, especially since we just got that cleaned up.

+1

File: impeller/toolkit/android/hardware_buffer.h
Comment (chinmaygarde) [reply impeller/toolkit/android/hardware_buffer.h]: I just put the size restrictions in the `@warning` block right above. Hoping to draw attention to it. But that was obviously not enough. I'll add a note reiterating the warning.

File: impeller/toolkit/android/native_window.h
Comment (chinmaygarde) [reply impeller/toolkit/android/native_window.h]: Done.

File: impeller/toolkit/android/hardware_buffer.cc
Comment (chinmaygarde) [reply impeller/toolkit/android/hardware_buffer.cc:82]: I documented that restriction in the headerdoc with a large warning sign so it shows up more prominently. It took me some trial and error to understand that this was the cause of buffers not being allocatable and the Android API doesn't say why a size wasn't allocatable. So making the callers figure this out was a pretty poor experience. Also, we do the same clamping in other spots (like the pbuffer sizes for offscreen GLES surfaces). So it seemed like in keeping with the theme. There are IsAllocatable checks either way though.

File: impeller/toolkit/android/surface_control.h
Comment (chinmaygarde) [reply impeller/toolkit/android/surface_control.h:22]: I am not aware of any threading restrictions on any of the handle types. The only major quirk is the transaction completion callback being made on a system managed thread (presumably a binder thread). I've documented that and also the lifecycle implications to the transaction because of it in the headerdoc comment for it.

Comment (chinmaygarde) [reply impeller/toolkit/android/surface_control.h:22]: I audited all calls, and beyond the completion callback that I have already documented, there is little guidance from the Android docs about the threading restrictions of any of the calls. I'm just going to be safe and call these as needing to be externally synchronized.

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (chinmaygarde) [reply impeller/toolkit/android/toolkit_android_unittests.cc:14]: > API works and that we can allocate/destroy these objects.

`CanCreateHardwareBuffer` for hardware buffers. `CanApplySurfaceTransaction` for surface transactions. Same as the NDK helpers.

> API symbols are correctly resolved.

`CanCreateProcTable`. You'll only get a valid proc table if the API symbols are correctly resolved and the version checks pass. For the individual subcomponents, this is also checked with `IsAvailableOnPlatform`.

> The tests here don't have as much coverage.

I'm surprised by this statement. There are only two tests in NDK helpers related to anything in this patch. `HardwareBuffer` and `SurfaceTransaction`. The entire corpus is only 5 tests in total ([unless I am missing something](https://github.com/flutter/engine/blob/main/fml/platform/android/ndk_helpers_unittests.cc#L79C1-L80C1)). The two tests I mentioned above more than cover the the assertions in NDK helpers.

I want to use the toolkit for everything and remove the NDK helpers after porting all existing tests. But, at least till I get to the choreographer, there isn't going to be anything to port as it should already be covered.

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (chinmaygarde) [reply impeller/toolkit/android/toolkit_android_unittests.cc:20]: I, uh, didn't want to link against `libAndroid.so`.

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (chinmaygarde) [reply impeller/toolkit/android/toolkit_android_unittests.cc:21]: Should I just do a GTEST_FAIL instead?

**Review by chinmaygarde — COMMENTED** [review]
> Do we intented to (eventually) replace the existing AHardwareBuffer and NDK wrappers used in the android embedder with this entirely?

Yes, I was going to replace the few uses of the helpers immediately after landing the swapchain patch and get rid of the libAndroid stuff in FML.

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (jonahwilliams) [reply impeller/toolkit/android/toolkit_android_unittests.cc:21]: GTEST_FAIL SGTM

Comment (dnfield) [reply impeller/toolkit/android/toolkit_android_unittests.cc:21]: We have infra that runs an API 28 emulator. 

We should probably have a test that verifies we have sane/non-crashing behavior if you use this API below 29 though...

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (dnfield) [reply impeller/toolkit/android/toolkit_android_unittests.cc:14]: We seem to be missing coverage here of `AHardwareBuffer_describe`, `AHardwareBuffer_isSupported`, and `AHardwareBuffer_getId`, which are needed by the current Android implementation.

Comment (chinmaygarde) [reply impeller/toolkit/android/toolkit_android_unittests.cc:14]: Yeah, but the toolkit has no ability to describe or get the ID of those objects yet. Once I work on adding those features get rid of NDK helpers, I can patch those in.

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (chinmaygarde) [reply impeller/toolkit/android/toolkit_android_unittests.cc:21]: > runs an API 28 emulator...

So, I can't make this a fail right?

> we have sane/non-crashing behavior if you use this API below 29 though

We are definitely missing a trick not using static analysis (like at-availability on iOS) for this. I spent a non-trivial amount of [time the other day wiring it up](https://discord.com/channels/608014603317936148/608021010377080866/1215804700046262334). But I haven't been able to get it to work reliably with our toolchain. Something to consider in addition to just checking on all emulator versions.

File: impeller/toolkit/android/README.md
Comment (chinmaygarde) [reply impeller/toolkit/android/README.md]: Done.

**Comment by chinmaygarde** [comment]
I believe I had addressed all open questions. PTAL. Past this, there are two patches to actually add the ABH swapchain implemnetation, and migrate existing users of NDK helpers. Then John should be able to wire up the Chroeographer VSync IDs.

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (dnfield) [reply impeller/toolkit/android/toolkit_android_unittests.cc:14]: The AHB code in ndk_helpers is used somewhere currently right? Even if not all of it.

Where is there at est in this file that creates a buffer, verfies its descriptor matches what it created, and verifies the buffer can be destroyed?

Comment (dnfield) [reply impeller/toolkit/android/toolkit_android_unittests.cc:14]: > Yeah, but the toolkit has no ability to describe or get the ID of those objects yet. Once I work on adding those features get rid of NDK helpers, I can patch those in.

I would rather see either Impeller use the FML utilities until it's ready to replace them with something better, or the rest of the code base using new improved utilities with similar test coverage. Otherwise we're likely to end up in a state where we have multiple locations to do NDK related lookups with different available/missing test coverage ...

Comment (chinmaygarde) [reply impeller/toolkit/android/toolkit_android_unittests.cc:14]: All I want to do is to facilitate landing https://github.com/flutter/engine/pull/51213. The ask on that patch was that it was too large and it should be split up. I have split it up into https://github.com/flutter/engine/pull/51295, https://github.com/flutter/engine/pull/51298, and this one. Two more patches are incoming after this to facilitate the request for smaller, more self-contained PRs.

I am on the hook for killing the FML utilities. I can kill the FML utilities in this patch if this gets us past the bikeshedding but it will make this patch bigger. Or, kill the FML utilities immediately after this lands if that seems more elegant.

Comment (chinmaygarde) [reply impeller/toolkit/android/toolkit_android_unittests.cc:14]: Ping @dnfield. Do you have a preference? Want to unblock this review ASAP.

Comment (dnfield) [reply impeller/toolkit/android/toolkit_android_unittests.cc:14]: I would rather eithe rkill it in this patch or update the AHB swapchain patch to use the FML utilities.

**Review by dnfield — CHANGES_REQUESTED** [review]
See https://github.com/flutter/engine/pull/51334/files#r1523948232

**Comment by chinmaygarde** [comment]
okie dokie. Killing the FML utils in this patch.

File: impeller/toolkit/android/proc_table.h
Comment (dnfield) [inline impeller/toolkit/android/proc_table.h:24]: We must not use this if `sizeof(uint64_t) != sizeof(long)`.

On 32 bit platforms we have to use the Java based API. But we still should try to resolve it on a 64 bit platform that is < API 29.  I made the same mistake when migrating things to the NDKHelpers class :)

Comment (dnfield) [reply impeller/toolkit/android/proc_table.h:24]: (That's why the `#if FML_ARCH_64BIT` guard was around this, or whatever it actually was)

**Comment by dnfield** [comment]
Overall this LGTM, will approve once the guard is added back to prevent undesirable usage of `AChoreographer_postFrameCallback(long` is back and CI is passing (looks like you missed at least one more `flutter::NDKHelpers` reference).

**Comment by chinmaygarde** [comment]
Oh, was still tinkering on this before signing off for the day. I doubt this compiles right now. Will ping again in the morning after the fixups.

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (chinmaygarde) [reply impeller/toolkit/android/toolkit_android_unittests.cc:21]: Oh, that's not why. `android_get_device_api_level` is itself only available on API level 29. So you can't call it on lower API levels. The act of checking if you are on a lower API level is itself unsafe.

**Review by chinmaygarde — COMMENTED** [review]
This is good to go. PTAL. Some updates since the last review:

* The NDK helpers are gone and existing users migrated.
* The toolkit now has a choreographer that handles the Callback32/64 variants under the hood.
* The previous version of the toolkit had an extra layer of protection over just the dysym. It checked not just that the API was available, but also that the platform was modern enough per the documentation. At the time, I was worried that an API may have been available as a symbol in an unstable form on older Android versions. But the way to check the Android API level is itself 29. The NDK helpers also didn't have the extra checks I had in the proc table. I have now removed the extra checks in the toolkit too. I had no reason to doubt this way isn't fine. Just need to trust the platform more.
  * Subsequently, the accessors to get the API level have also been removed.

File: impeller/toolkit/android/toolkit_android_unittests.cc
Comment (chinmaygarde) [inline impeller/toolkit/android/toolkit_android_unittests.cc:95]: @dnfield @jonahwilliams This test will currently hang because the unit-test run as a naked binary don't get choreographer callbacks. I'll remove this skip once I have the APK based unit-tests harness we discussed yesterday.

File: impeller/toolkit/android/proc_table.cc
Comment (dnfield) [reply impeller/toolkit/android/proc_table.cc:50]: SGTM

**Review by dnfield — APPROVED** [review]
LGTM as long as CI is happy!

**Comment by matanlurey** [comment]
Reason for revert: Broke engine post-submit, see https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8753367119442265873/+/u/test:_Android_Unit_Tests__API_28_/stdout.

**Pull request merged**
